src/python/pants/base/specs.py

```diff
@@ -147,43 +139,48 @@ class SiblingAddresses(AddressSpec):
     def __str__(self) -> str:
         return f"{self.directory}:"
 
-    def matching_address_families(self, address_families_dict: Dict[str, Any]) -> List[Any]:
-        return self.address_families_for_dir(address_families_dict, self.directory)
+    def to_globs(self, build_patterns: Iterable[str]) -> Tuple[str, ...]:
+        return _globs_in_single_dir(self.directory, build_patterns)
 
-    def address_target_pairs_from_address_families(self, address_families: Sequence[Any]):
-        return self.all_address_target_pairs(address_families)
+    def matching_address_families(
+        self, address_families_dict: Mapping[str, "AddressFamily"]
+    ) -> Tuple["AddressFamily", ...]:
+        return (_address_family_for_dir(self.directory, address_families_dict),)
 
-    def to_globs(self, build_patterns: Iterable[str]) -> Tuple[str, ...]:
-        return self.globs_in_single_dir(self.directory, build_patterns)
+    def matching_addresses(
+        self, address_families: Sequence["AddressFamily"]
+    ) -> Sequence[Tuple[Address, TargetAdaptor]]:
+        return _all_address_target_pairs(address_families)
 
 
 @dataclass(frozen=True)
 class DescendantAddresses(AddressSpec):
     """An AddressSpec representing all addresses located recursively under the given directory."""
 
     directory: str
-    error_if_no_matches: bool = True
 
     def __str__(self) -> str:
         return f"{self.directory}::"
 
-    def matching_address_families(self, address_families_dict: Dict[str, Any]) -> List[Any]:
-        return [
+    def to_globs(self, build_patterns: Iterable[str]) -> Tuple[str, ...]:
+        return tuple(os.path.join(self.directory, "**", pat) for pat in build_patterns)
+
+    def matching_address_families(
+        self, address_families_dict: Mapping[str, "AddressFamily"]
+    ) -> Tuple["AddressFamily", ...]:
+        return tuple(
             af
             for ns, af in address_families_dict.items()
             if fast_relpath_optional(ns, self.directory) is not None
-        ]
-
-    def address_target_pairs_from_address_families(self, address_families: Sequence[Any]):
-        addr_tgt_pairs = self.all_address_target_pairs(address_families)
-        if self.error_if_no_matches and len(addr_tgt_pairs) == 0:
-            raise self.AddressResolutionError(
-                f"Address spec '{str(self)}' does not match any targets."
-            )
-        return addr_tgt_pairs
+        )
 
-    def to_globs(self, build_patterns: Iterable[str]) -> Tuple[str, ...]:
-        return tuple(os.path.join(self.directory, "**", pat) for pat in build_patterns)
+    def matching_addresses(
+        self, address_families: Sequence["AddressFamily"]
+    ) -> Sequence[Tuple[Address, TargetAdaptor]]:
+        matching = _all_address_target_pairs(address_families)
+        if len(matching) == 0:
+            raise ResolveError(f"Address spec '{str(self)}' does not match any targets.")
```

Comment:
nit: `SiblingAddresses` will only error if no BUILD files existed in the relevant directories (ie, not if there will BUILD files but they were empty)... should `DescendantAddresses` do the same?

(fine to ignore this too: I think it's mostly that's this change makes it much easier to tell what's going on!)

Reply:
On second thought, going to punt on this. I'm not sure what the correct behavior should be for either type of spec.